### DISCUSSION TITLE
[@types/react-native] StyleSheet.create won't raise error for unexisting properties

### DISCUSSION THREAD

**Discussion by xsjcTony** [DefinitelyTyped/DefinitelyTyped#62481] [Issues with a @types package]
Basically, TypeScript won't raise any Error for the object passed into `StyleSheet.create()` with any kind of unexisting properties.
But I have totally no idea why.
I didn't find anything like `[K: string]: AnyType` in `ViewStyle | TextStyle | ImageStyle`, which is the type of all available styles.

The key is, does `extends` in generic mean `assignable` ?

[Link to CodeSandbox](https://codesandbox.io/s/react-native-typescript-issue-u2d8kq?file=/index.ts)

Code is also available here
```TypeScript
import { StyleSheet } from "react-native";
import type { StyleSheet as SS } from "react-native";


// Original signature of StyleSheet.create
// https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f6282734f8d04a3653d6f61d2a37691107d64845/types/react-native/index.d.ts#L5465
declare function create<T extends SS.NamedStyles<any> | SS.NamedStyles<T>>(
  styles: SS.NamedStyles<T> | T
): T;

// Considered as valid when using StyleSheet.create (inferring generic type), but why?
// `extends` keyword in generic means "assignable" to me, but it's not assignable in this case, proved down there (even with any).
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",

    // These should raise errors, but why is it considered as valid?
    random: 123,
    nested: {
      width: 200,
      height: 70,
      backgroundColor: "dodgerblue"
    }
  }
});

// Error is raised when assigning straight away
const obj: SS.NamedStyles<any> = {
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",

    // Error raised as expected
    random: 123,
    nested: {
      width: 200,
      height: 70,
      backgroundColor: "dodgerblue"
    }
  }
};
```

And surprisingly the nested style works fine `style={styles.container.nested}`

---

I have no idea why my react-native app is not raising runtime errors when misspelled style property, and TypeScript is not checking that either.

[a comment by an automated account is not shown]

**Comment by xsjcTony** [comment]
Any help? :(

**Comment by RobJacobson** [comment]
I'm encountering this problem too. I created a new project using Expo (with TypeScript), and even uninstalled/reinstalled VS Code without extensions as a sanity check. No luck.

Specifically, VS Code/TypeScript _works correctly_ and shows a "red squiggly" to indicate an error in either of these circumstances:

1. If I define a style inline, e.g.:

`export default function App() {
  return (
    <View
      style={{
        flex: 1,
        backgroundColorz: "#ddd",
      }}
    >
      <Text>Open up App.js to start working on your app!</Text>
    </View>
  );
}`

2. If I define a style object as a constant, defined with the ViewStyle type:

`export default function App() {
  return (
    <View style={container}>
      <Text>Open up App.js to start working on your app!</Text>
    </View>
  );
}

const container: ViewStyle = {
  flex: 1,
  backgroundColorz: "#ddd",
};
`

However, if I attempt to use StyleSheet.create(), VSCode _does not show_ any error. It accepts this code without complaining, both at development time and at build/run time:

`export default function App() {
  return (
    <View
      style={{
        flex: 1,
        backgroundColorz: "#ddd",
      }}
    >
      <Text>Open up App.js to start working on your app!</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColorr: "#fff",
  },
});`

As a weird exception, VS Code _does show_ an error (correctly) if I call StyleSheet.create() with just a single style definition:

`export default function App() {
  return (
    <View style={styles.container}>
      <Text>Open up App.js to start working on your app!</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    backgroundColorr: "#fff",
  },
});`

In this case only, VS Code shows a "red squiggly" underneath the incorrect "backgroundColorr" definition in the last line. This seems like an important clue, but I don't understand the reason.

I really hope this can be fixed. I find it distressing that TypeScript would "allow" this error without informing me. If I make a typo, VS Code will gladly accept it.

In the interim, for both me and @xsjcTony, it seems like using the `const container: ViewStyle = {...}` alternative works OK. From the links below, there doesn't seem to be any compelling reason for using StyleSheet.create() over creating individual style constants.

https://stackoverflow.com/questions/38958888/react-native-what-is-the-benefit-of-using-stylesheet-vs-a-plain-object

https://github.com/infinitered/ignite/blob/master/docs/Styling.md

(Edit: Sorry, I don't know why the code isn't formatting correctly.)

**Comment by krosenberg** [comment]
I ran into this issue too, and then went down a two hour rabbit hole to figure out what was going on.

This isn't related to the type definitions for React Native -- but rather how Typescript handles [Excess Property Checks](https://www.typescriptlang.org/docs/handbook/2/objects.html#excess-property-checks).

tl;dr -- unless you assign a type to an object _at the same point where the object_ is created, Typescript won't check for excess types. One exception: if the object you're passing has only "excess" types, it _will_ catch that.

So in this [example](https://www.typescriptlang.org/play?ts=5.1.3&ssl=16&ssc=1&pln=2&pc=1#code/FASwdgLgpgTgZgQwMZQAQEEwgLYIDYDOqA3qgCYD2A5gPwBcqBEM4VqAvsHAK5hIQgKYVAAtuVTDnwEAFAiy5CDSYoIBKEp2DAA9DtQA5CqlgwKMYEiFNU8qYVQBeEpSoMA5ACEEMANYBPdwAaVCQECA8AWSgKAHdAjgBuYDEJBWk5dMI1ZN19AFEYMxgCS2sIWyyCACZlKqcXag9vP0CQsIjUd2i4hPZk1JUMu1VqnO09VELi0sGqmWJXZp8A4NDwqJj49w5xoA):
```ts
interface Animals { dog?: string }
function hugAnimals(animals: Animals) {}
```

You'd get the following results:
```ts
// No error -- animals doesn't get a type def at creation
const animals = {dog: 'Barky', cat: 'Meowy' };
hugAnimals(animals);

// Error -- animals2 gets a type def (and type checked) at creation
const animals2: Animals = {dog: 'Barky', cat: 'Meowy' };
hugAnimals(animals2);

// Error -- the object is getting a type def because it's created within the function call, and the function's argument has a type def
hugAnimals({dog: 'Barky', cat: 'Meowy' });
```

This was surprising to me as someone who used to work a lot with Flow but is newer to Typescript, where "exact types" (used to be done with `{||}` before it became the default) would prevent this type of thing.

**Reply by xsjcTony** [reply]
But I'm passing the object into `StyleSheet.create()` straight away, which should match the third case there. It still not giving me errors
